File: csharp/Maglev.cs
namespace ConsistentHash;

/// <summary>
/// Maglev lookup-table hashing (Google, 2016). A fixed prime-sized permutation table
/// is built once; each node claims slots in a deterministic interleaving. Lookups are
/// a single array index, and losing a node only rewrites its own slots.
/// </summary>
public sealed class Maglev
{
    private readonly int _m;
    private List<string> _nodes;
    private int[] _table = Array.Empty<int>();

    public Maglev(IEnumerable<string>? nodes = null, int tableSize = 65537)
    {
        if (!IsPrime(tableSize))
        {
            throw new ArgumentException("tableSize must be prime", nameof(tableSize));
        }

        _m = tableSize;
        _nodes = (nodes ?? Enumerable.Empty<string>())
            .Distinct()
            .OrderBy(n => n, StringComparer.Ordinal)
            .ToList();
        Build();
    }

    public void Add(string node)
    {
        if (!_nodes.Contains(node))
        {
            _nodes.Add(node);
            _nodes = _nodes.OrderBy(n => n, StringComparer.Ordinal).ToList();
            Build();
        }
    }

    public void Remove(string node)
    {
        if (_nodes.Remove(node))
        {
            Build();
        }
    }

    public string? Get(string key)
    {
        if (_nodes.Count == 0)
        {
            return null;
        }

        return _nodes[_table[HashRing.Fnv1a64(key) % (ulong)_m]];
    }

    public IReadOnlyList<string> Nodes => _nodes.ToList();

    public int Count => _nodes.Count;

    private void Build()
    {
        var n = _nodes.Count;
        if (n == 0)
        {
            _table = Array.Empty<int>();
            return;
        }

        var offset = new ulong[n];
        var skip = new ulong[n];
        for (var i = 0; i < n; i++)
        {
            offset[i] = HashRing.Fnv1a64($"{_nodes[i]}#offset") % (ulong)_m;
            skip[i] = (HashRing.Fnv1a64($"{_nodes[i]}#skip") % (ulong)(_m - 1)) + 1;
        }

        var table = new int[_m];
        Array.Fill(table, -1);
        var next = new ulong[n];
        var filled = 0;
        while (filled < _m)
        {
            for (var i = 0; i < n; i++)
            {
                var c = (offset[i] + (next[i] * skip[i])) % (ulong)_m;
                while (table[c] >= 0)
                {
                    next[i]++;
                    c = (offset[i] + (next[i] * skip[i])) % (ulong)_m;
                }

                table[c] = i;
                next[i]++;
                filled++;
                if (filled == _m)
                {
                    break;
                }
            }
        }

        _table = table;
    }

    private static bool IsPrime(int n)
    {
        if (n < 2)
        {
            return false;
        }

        if (n % 2 == 0)
        {
            return n == 2;
        }

        for (var i = 3; i * i <= n; i += 2)
        {
            if (n % i == 0)
            {
                return false;
            }
        }

        return true;
    }
}

File: go/maglev.go
package consistenthash

import (
	"fmt"
	"sort"
)

// Maglev is Google's lookup-table hashing: build a fixed prime-sized permutation
// table once, where each node claims slots in a deterministic interleaving. Lookups
// are a single array index, balance is tighter than the ring, and losing a node only
// rewrites its own slots.
type Maglev struct {
	m     int
	nodes []string
	table []int
}

// NewMaglev builds a Maglev table of the given prime size over nodes.
func NewMaglev(nodes []string, tableSize int) *Maglev {
	if !isPrime(tableSize) {
		panic("tableSize must be prime")
	}
	m := &Maglev{m: tableSize, nodes: dedupeSorted(nodes)}
	m.build()
	return m
}

// Add inserts a node and rebuilds the table.
func (m *Maglev) Add(node string) {
	for _, n := range m.nodes {
		if n == node {
			return
		}
	}
	m.nodes = append(m.nodes, node)
	sort.Strings(m.nodes)
	m.build()
}

// Remove deletes a node and rebuilds the table.
func (m *Maglev) Remove(node string) {
	for i, n := range m.nodes {
		if n == node {
			m.nodes = append(m.nodes[:i], m.nodes[i+1:]...)
			m.build()
			return
		}
	}
}

// Get returns the node owning key, or "" if empty.
func (m *Maglev) Get(key string) string {
	if len(m.nodes) == 0 {
		return ""
	}
	return m.nodes[m.table[Fnv1a64(key)%uint64(m.m)]]
}

// Nodes returns the nodes in sorted order.
func (m *Maglev) Nodes() []string {
	out := make([]string, len(m.nodes))
	copy(out, m.nodes)
	return out
}

// Len is the number of nodes.
func (m *Maglev) Len() int { return len(m.nodes) }

func (m *Maglev) build() {
	n := len(m.nodes)
	if n == 0 {
		m.table = nil
		return
	}
	offset := make([]uint64, n)
	skip := make([]uint64, n)
	for i, name := range m.nodes {
		offset[i] = Fnv1a64(fmt.Sprintf("%s#offset", name)) % uint64(m.m)
		skip[i] = Fnv1a64(fmt.Sprintf("%s#skip", name))%uint64(m.m-1) + 1
	}
	table := make([]int, m.m)
	for i := range table {
		table[i] = -1
	}
	next := make([]uint64, n)
	filled := 0
	for filled < m.m {
		for i := 0; i < n; i++ {
			c := (offset[i] + next[i]*skip[i]) % uint64(m.m)
			for table[c] >= 0 {
				next[i]++
				c = (offset[i] + next[i]*skip[i]) % uint64(m.m)
			}
			table[c] = i
			next[i]++
			filled++
			if filled == m.m {
				break
			}
		}
	}
	m.table = table
}

func isPrime(n int) bool {
	if n < 2 {
		return false
	}
	if n%2 == 0 {
		return n == 2
	}
	for i := 3; i*i <= n; i += 2 {
		if n%i == 0 {
			return false
		}
	}
	return true
}

func dedupeSorted(nodes []string) []string {
	seen := make(map[string]struct{}, len(nodes))
	out := make([]string, 0, len(nodes))
	for _, n := range nodes {
		if _, ok := seen[n]; !ok {
			seen[n] = struct{}{}
			out = append(out, n)
		}
	}
	sort.Strings(out)
	return out
}
